docs: service anatomy (#38)
equinox-project/equinox-js: docs/docs/anatomy.md

---
sidebar_position: 3
---

# Anatomy of an Equinox Service

An Equinox service is composed of a few modules (namespaces).

- **Stream**: tells us which category the service writes to and how the identity

>   of the stream is composed

Reviewer: ```suggestion
  of the stream is composed (and the reverse operations for when running Reaction logic)
```

> - **Events**: tells us which events the service writes

Reviewer: and how they are to be mapped to/from the stored form (e.g. applying wrapper types over primitives)

but only if you an convey it more tersely
the aim being to sell why they are needed and not overkill in a one liner

- **Fold**: tells us how the events are folded into state
- **Decide**: tells us which actions can be taken upon the state (resulting in
  new events being written)

> - **Service**: the class that wraps the above into a cohesive domain service.

Reviewer: maybe a `Queries` section for rendering helpers and dto types like https://github.com/jet/equinox/pull/429 or is this too early?

Author: Definitely not too early, I'll add it and mention it as optional

Reviewer: also maybe call out that the other things are private/pubternal and this is the thing that you talk to

Author: > pubternal

Only to be discussed at the pub, never in the office

Author: Query section added

> - **Config**: tells us how to wire the service against concrete stores,

Reviewer: ```suggestion
- **Config**: shows how we've opted to to bind the Service to Streams in a Concrete Store (which Access Strategies to apply, if any)
```

  including important information like access strategies.

Let's see what each of these modules looks like for a simple service for

> checking in and out of an appointment.

Reviewer: ```suggestion
checking in and out for an Appointment Booking.
```

  // prettier-ignore
  const date = z.string().datetime().transform((x) => new Date(x))
  export const Timestamp = z.object({ timestamp: date })
  export const ActualsOverridden = z.object({ checkedIn: date, checkedOut: date })
  export type ActualsOverridden = z.infer<typeof ActualsOverridden>

  export type Event =
    | { type: "CheckedIn"; data: Timestamp }
    | { type: "CheckedOut"; data: Timestamp }
    | { type: "ActualsOverridden"; data: ActualsOverridden }

  export const codec = Codec.upcast<Event>(
    Codec.json(),
    Codec.Upcast.from({
      CheckedIn: Timstamp.parse,
      CheckedOut: Timstamp.parse,
      ActualsOverridden: ActualsOverridden.parse,
    }),
  )
}
```

## The Fold

```ts
export namespace Fold {
  import Event = Events.Event
  export type State = { checkedIn?: Date; checkedOut?: Date }
  export const initial: State = {}
  export const evolve = (state: State, event: Event) => {
    switch (event.type) {
      case "CheckedIn":
        return { ...state, checkedIn: event.data.timestamp }
      case "CheckedOut":
        return { ...state, checkedOut: event.data.timestamp }
      case "ActualsOverridden":
        return event.data
    }
  }
}
```

> ## The Decide

Reviewer: this reads wacko
And while its perfect for a small example, it odoes not show how you might subdivide it for multiple decisions

Author: This'll be the second item in the nav. Not sure if it's the right place for a full-blown complex thing. Don't want to scare people just yet 😅

```ts
export namespace Decide {
  import Event = Events.Event
  import State = Fold.State
  export const checkIn =
    (timestamp: Date) =>
    (state: State): Event[] => {

>       if (state.checkedIn) {

Reviewer: while this is subjective and has pros
I think 
1. doing a check on the conjunction and throwing
2. then having an if state.checkedIn vs not to illustrate silent idempotent handling as the main thing
3. then the happy path
might convey the rules of the game better (also `no change` conveyed it to me, but its a bit subtle and might merit something more long winded like saying 'caller is expected to guard against this' and `we accept retried requests (with identical timestamps)`)

        // no change
        if (+state.checkedIn === +timestamp) return []
        throw new Error("Already checked in with different timestamp")
      }
      return [{ type: "CheckedIn", data: { timestamp } }]
    }
  export const checkOut =
    (timestamp: Date) =>
    (state: State): Event[] => {
      if (state.checkedOut) {
        // no change
        if (+state.checkedOut === +timestamp) return []
        throw new Error("Already checked out with different timestamp")
      }
      return [{ type: "CheckedOut", data: { timestamp } }]
    }

  export const manuallyOverride = (checkedIn: Date, checkedOut: Date) => (state: State) => {

>     if (+checkedIn === +state?.checkedIn && +checkedOut === +state?.checkedIn) return []

Reviewer: if there are comments on the above explicatting the concept, this is a fantastic example

    return [{ type: "ActualsOverridden", data: { checkedIn, checkedOut } }]
  }
}
```

## The Service

```ts
export class Service {
  constructor(
    private readonly resolve: (
      appointmentId: AppointmentId,
      userId: UserId,
    ) => Decider<Events.Event, Fold.State>,
  ) {}

  checkIn(appointmentId: AppointmentId, userId: UserId, timestamp: Date) {
    const decider = this.resolve(appointmentId, userId)
    return decider.transact(Decide.checkIn(timestamp), LoadOption.AssumeEmpty)
  }

  checkOut(appointmentId: AppointmentId, userId: UserId, timestamp: Date) {
    const decider = this.resolve(appointmentId, userId)

>     return decider.transact(Decide.checkOut(timestamp), LoadOption.AssumeEmpty)

Reviewer: dont assume empty for checkOut
a) coz it wont be
b) to underscore that most of the time you dont do or think about fancy options

Author: ```suggestion
    return decider.transact(Decide.checkOut(timestamp))
```

Reviewer: yes, think so

  }

  manuallyOverride(
    appointmentId: AppointmentId,
    userId: UserId,
    checkedIn: Date,
    checkedOut: Date,
  ) {
    const decider = this.resolve(appointmentId, userId)
    return decider.transact(Decide.manuallyOverride(checkedIn, checkedOut))
  }
}
```

## The Config

<details>
<summary>See the `../equinox-config.ts` snippet</summary>

```ts
import { ICodec, ICache, CachingStrategy, Codec } from "@equinox-js/core"
import { MemoryStoreCategory, VolatileStore } from "@equinox-js/memory-store"
import * as MessageDB from "@equinox-js/message-db"
import * as DynamoDB from "@equinox-js/dynamo-store"

export enum Store {
  Memory,
  MessageDb,
  Dynamo,
}

export type Config =
  | { store: Store.Memory; context: VolatileStore<string> }
  | { store: Store.MessageDb; context: MessageDB.MessageDbContext; cache: ICache }
  | { store: Store.Dynamo; context: DynamoDB.DynamoStoreContext; cache: ICache }

// prettier-ignore
export namespace MessageDb {
  import AccessStrategy = MessageDB.AccessStrategy

>   import MessageDbCategory = MessageDB.MessageDbCategory

Reviewer: short nams like L195?

    const access = AccessStrategy.LatestKnownEvent<E, S>()
    return MessageDb.createCached(name, codec, fold, initial, access, config)
  }
}

// prettier-ignore
export namespace Dynamo {
  import AccessStrategy = DynamoDB.AccessStrategy
  import Category = DynamoDB.DynamoStoreCategory
  import Context = DynamoDB.DynamoStoreContext
  type Config = { context: Context; cache: ICache }
  export function createCached<E, S, C>(name: string, codec_: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, access: AccessStrategy<E, S>, { context, cache }: Config) {
    const caching = CachingStrategy.Cache(cache)
    const codec = Codec.compress(codec_)
    return Category.create(context, name, codec, fold, initial, caching, access);
  }
  export function createUnoptimized<E, S, C>(name: string, codec: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, config: Config) {
    const access = AccessStrategy.Unoptimized()
    return Dynamo.createCached(name, codec, fold, initial, access, config)
  }

>   export function createLatestKnown<E,S,C>(name: string, codec: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, config: Config) {

Reviewer: this is last in mdb - prob last here too as pretty rare

Author: Yeah, definitely more common to use snapshots than latest event in dynamo. In MDB I've got more services on LatestKnownEvent than AdjacentSnapshots.

  }
  export function createSnapshotted<E,S,C>(name: string, codec: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, isOrigin: (e: E) => boolean, toSnapshot: (s: S) => E, config: Config) {
    const access = AccessStrategy.Snapshot(isOrigin, toSnapshot)
    return Dynamo.createCached(name, codec, fold, initial, access, config)
  }
  export function createRollingState<E,S,C>(name: string, codec: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, toSnapshot: (s: S) => E, config: Config) {
    const access = AccessStrategy.RollingState(toSnapshot)
    return Dynamo.createCached(name, codec, fold, initial, access, config)
  }
}

// prettier-ignore
export namespace MemoryStore {
  export function create<E, S, C>(name: string, codec: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, { context: store }: { context: VolatileStore<string> }) {
    return MemoryStoreCategory.create(store, name, codec, fold, initial)
  }
}
```

</details>

```ts
import * as Config from "../equinox-config"
class Service {
  // ... as above

  // prettier-ignore
  static resolveCategory(config: Config.Config) {
    switch (config.store) {
      case Config.Store.Memory:

>         return Config.MemoryStore.create(Stream.name, Events.codec, Fold.fold, Fold.initial, config)

Reviewer: ```suggestion
        return Config.MemoryStore.create(Stream.category, Events.codec, Fold.fold, Fold.initial, config)
```

      case Config.Store.MessageDB:
        return Config.MessageDB.createUnoptimized(Stream.name, Events.codec, Fold.fold, Fold.initial, config)
      case Config.Store.Dynamo:
        return Config.Dynamo.createUnoptimized(Stream.name, Events.codec, Fold.fold, Fold.initial, config)
    }
  }
  static create(config: Config.Config) {
    const category = Service.resolveCategory(config)
    const resolve = (appointmentId: AppointmentId, userId: UserId) =>

>       Decider.forStream(category, Stream.streamId(appointmentId, userId), null)

Reviewer: name the null and/or leave it optional?

Author: ATM the null is necessary, will be making it optional soon 👍

Author: see #41

An Equinox service is composed of a few modules (namespaces).

- **Stream**: shows which category the service writes to and how the identity
  of the stream is composed (and the reverse operations for when running Reaction logic)
- **Events**: shows which events the service writes
- **Fold**: shows how the events are folded into state
- **Decide**: shows which actions can be taken upon the state (resulting in
  new events being written)
- **Service**: the class that wraps the above into a cohesive domain service.
- **Config**: shows how we've opted to to bind the Service to Streams in a Concrete Store (which Access Strategies to apply, if any)
  including important information like access strategies.

:::tip

Everything above `Service` is considered internal to the module. It is exported

> as a conveninence for testing, but other modules should not take a dependency on

Reviewer: ```suggestion
as a convenience for testing, but other modules should not take a dependency on
```

<details>
<summary>See the `../equinox-config.ts` snippet</summary>

> Feel free to copy-paste this snippet for your own project.

Reviewer: Maybe mention that in general you have one set of this binding per repo ?
And that in most real apps you'll typically have bindings for one concrete store and the memory store

Author: modified

// prettier-ignore
export namespace MessageDb {

>   import AccessStrategy = MessageDB.AccessStrategy

Reviewer: hm should AccessStrategy be prefixed too

Author: I'd rather the other stuff lost its prefix

```ts
DynamoStore.DynamoStoreCategory.create(....)
// vs
DynamoStore.Category.create(...)
```

Reviewer: yeah, but some names are very generic (client, category, context)
and then there are other conventional names like Config, Store, Streams also in the same scope
Given the names should only really be used in the context of one of those Config binding wiring blocks, the fact they are a bit long is no real harm
Also Category would clash with Equinox.Category (and I wanted to rename ISyncContext to Context but again same concerns stop me esp given how overloaded the term Context is)

A lot of these things get close enough to files that have domain logic too - things like Category etc can easily encroach on and/or require constant disambiguation.

Can't see where I saw the argument for the long names - it was in some Azure Core guidlines or sometihng (not they should be conodered arbiters of taste for even 5 mins!)

  import Category = MessageDB.MessageDbCategory
  import Context = MessageDB.MessageDbContext
  type Config = { context: Context; cache: ICache }

  export function createCached<E, S, C>(name: string, codec: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, access: AccessStrategy<E, S>, { context, cache }: Config) {
    const caching = CachingStrategy.Cache(cache)
    return Category.create(context, name, codec, fold, initial, caching, access);
  }
  export function createUnoptimized<E, S, C>(name: string, codec: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, config: Config) {
    const access = AccessStrategy.Unoptimized<E, S>()
    return MessageDb.createCached(name, codec, fold, initial, access, config)
  }
  export function createSnapshotted<E, S, C>(name: string, codec: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, eventName: string, toSnapshot: (s: S) => E, config: Config) {
    const access = AccessStrategy.AdjacentSnapshots(eventName, toSnapshot)
    return MessageDb.createCached(name, codec, fold, initial, access, config)
  }

>   export function createLatestKnown<E, S, C>(name: string, codec: ICodec<E, string, C>, fold: (s: S, e: E[]) => S, initial: S, config: Config) {

Reviewer: latesknown is not last for Dynamo

Author: It is now

  static resolveCategory(config: Config.Config) {
    switch (config.store) {
      case Config.Store.Memory:

>         return Config.MemoryStore.create(Stream.category, Events.codec, Fold.fold, Fold.initial, config)

Reviewer: see clones
